pandablocks/ioc/ioc.py

```diff
@@ -0,0 +1,1756 @@
+        self,
+        record_name: EpicsName,
+        field_info: FieldInfo,
+        values: Dict[EpicsName, ScalarRecordValue],
+    ) -> Dict[EpicsName, RecordInfo]:
+        return self._make_enum(record_name, field_info, values, builder.mbbIn)
+
+    def _make_enum_write(
+        self,
+        record_name: EpicsName,
+        field_info: FieldInfo,
+        values: Dict[EpicsName, ScalarRecordValue],
+    ) -> Dict[EpicsName, RecordInfo]:
+        assert isinstance(field_info, EnumFieldInfo)
+        assert field_info.labels
+        assert record_name not in values
+        # Values are not returned for write fields. Create data for label parsing.
+        values = {record_name: field_info.labels[0]}
+        return self._make_enum(
+            record_name, field_info, values, builder.mbbOut, always_update=True
+        )
+
+    def create_record(
+        self,
+        record_name: EpicsName,
+        field_info: FieldInfo,
+        field_values: Dict[EpicsName, RecordValue],
+    ) -> Dict[EpicsName, RecordInfo]:
+        """Create the record (and any child records) for the PandA field specified in
+        the parameters.
+
+        Args:
+            record_name: The name of the record to create, with colons separating
+                words in EPICS style.
+            field_info: The field info for the record being created
+            field_values: The dictionary of values for the record and
+                all child records. The keys are in EPICS style.
+
+        Returns:
+            Dict[str, RecordInfo]: A dictionary of created RecordInfo objects that
+                will need updating later. Not all created records are returned.
+        """
+
+        try:
+            key = (field_info.type, field_info.subtype)
+            if key == ("table", None):
+                # Table expects vals in Dict[str, List[str]]
+                list_vals = {
+                    EpicsName(k): v
+                    for (k, v) in field_values.items()
+                    if isinstance(v, list)
+                }
+
+                return self._make_table(record_name, field_info, list_vals)
+
+            # PandA can never report a table field as in error, only scalar fields
+            str_vals = {
+                EpicsName(k): v
+                for (k, v) in field_values.items()
+                if isinstance(v, (str, InErrorException))
+            }
+
+            return self._field_record_mapping[key](
+                self, record_name, field_info, str_vals
+            )
+
+        except KeyError:
+            # Unrecognised type-subtype key, ignore this item. This allows the server
+            # to define new types without breaking the client.
+            logging.exception(
+                f"Unrecognised type {key} while processing record {record_name}"
+            )
+            return {}
+
+    # Map a field's (type, subtype) to a function that creates and returns record(s)
+    _field_record_mapping: Dict[
```

Comment:
Could you avoid a layer of `_make` functions if you put a few more things in this table? Like record factory, or number of values?

Reply:
Probably possible, I'll investigate...